// bcryptjs/index.js
const mongoose = require("mongoose");
const bcrypt = require("bcryptjs");

const userSchema = new mongoose.Schema(
  {
    // 👤 Basic Account Info
    name: {
      type: String,
      required: [true, "Name is required"],
      trim: true,
    },
    email: {
      type: String,
      required: [true, "Email is required"],
      unique: true,
      lowercase: true,
      match: [/^\S+@\S+\.\S+$/, "Please use a valid email address"],
    },
    password: {
      type: String,
      required: [true, "Password is required"],
      minlength: 6,
      select: false, // hide password by default
    },

    // 🌿 Extended Profile Fields
    age: { type: Number, min: 0, max: 120, default: null },
    gender: { type: String, enum: ["Male", "Female", "Other", ""], default: "" },
    phone: {
      type: String,
      trim: true,
      match: [/^[0-9]{10}$/, "Please enter a valid 10-digit phone number"],
      default: "",
    },
    location: { type: String, trim: true, maxlength: 100, default: "" },
    goal: { type: String, trim: true, maxlength: 100, default: "" },
    bio: { type: String, trim: true, maxlength: 300, default: "" },
    profilePic: { type: String, default: "" },
    isProfileComplete: { type: Boolean, default: false },
  },
  { timestamps: true }
);

// 🧠 Hash password before saving
userSchema.pre("save", async function (next) {
  if (!this.isModified("password")) return next();
  try {
    const salt = await bcrypt.genSalt(10);
    this.password = await bcrypt.hash(this.password, salt);
    next();
  } catch (err) {
    next(err);
  }
});

// 🔑 Compare entered password with stored hash
userSchema.methods.matchPassword = async function (enteredPassword) {
  return bcrypt.compare(enteredPassword, this.password);
};

// ✅ Hide sensitive fields
userSchema.methods.toJSON = function () {
  const obj = this.toObject();
  delete obj.password;
  delete obj.__v;
  return obj;
};

module.exports = mongoose.model("User", userSchema);
